readme.md
Quant
=====

Interface for color quantizers.


changelog
v0.1 2013 Sep 20

This started as a little toy program to implement color quantization.  But then
I looked to see what else was published in Go along these lines and didn't find
much.  To add something of interest, I implemented a second algorithm and added
an interface.  That's about all that is in v0.1 here.  It's far from general
utility.  It needs things like dithering, subsampling, optimization for common
image types, and test code.


quant.go
// Copyright 2013 Sonia Keys.
// Licensed under MIT license.  See "license" file in this source tree.

// Quant provides an interface for image color quantizers.
package quant

import (
	"image"
	"image/color"
)

// Quantizer defines a color quantizer for images.
type Quantizer interface {
	// Quantize int argument specifies the desired number of colors
	// in the result image.
	Image(image.Image, int) *image.Paletted
	Palette(image.Image, int) Palette
}

type Palette interface {
	Convert(color.Color) color.Color
	Index(color.Color) int
	ColorPalette() color.Palette
}

type LinearPalette struct {
	color.Palette
}

func (p LinearPalette) ColorPalette() color.Palette {
	return p.Palette
}

func Dither(i0 image.Image, p Palette) *image.Paletted {
	cp := p.ColorPalette()
	if len(cp) > 256 {
		return nil
	}
	b := i0.Bounds()
	pi := image.NewPaletted(b, cp)
	if b.Max.Y-b.Min.Y == 0 || b.Max.X-b.Min.X == 0 {
		return pi
	}
	// rt, dn hold diffused errors.
	var rt color.RGBA64
	dn := make([]color.RGBA64, b.Max.X - b.Min.X + 1)
	for y := b.Min.Y; y < b.Max.Y; y++ {
		rt = dn[0]
		dn[0] = color.RGBA64{}
		for x := b.Min.X; x < b.Max.X; x++ {
			// full color from original image
			c0 := i0.At(x, y)
			r0, g0, b0, _ := c0.RGBA()
			// adjusted full color = original color + diffused error
			r0 += uint32(rt.R)
			g0 += uint32(rt.G)
			b0 += uint32(rt.B)
			if r0 > 0xffff {
				r0 = 0xffff
			}
			if g0 > 0xffff {
				g0 = 0xffff
			}
			if b0 > 0xffff {
				b0 = 0xffff
			}
			rt.R = uint16(r0)
			rt.G = uint16(g0)
			rt.B = uint16(b0)
			// nearest palette entry
			i := cp.Index(rt)
			pi.SetColorIndex(x, y, uint8(i))
			pr, pg, pb, _ := cp[i].RGBA()
			// error to be diffused = full color - palette color.
			// half of error goes right
			if uint16(pr) > rt.R {
				rt.R = 0
			} else {
				rt.R = (rt.R-uint16(pr))/2
			}
			if uint16(pg) > rt.G {
				rt.G = 0
			} else {
				rt.G = (rt.G-uint16(pg))/2
			}
			if uint16(pb) > rt.B {
				rt.B = 0
			} else {
				rt.B = (rt.B-uint16(pb))/2
			}
			// half goes down
			dn[x+1].R = rt.R/2
			dn[x+1].G = rt.G/2
			dn[x+1].B = rt.B/2
			dn[x].R += dn[x+1].R
			dn[x].G += dn[x+1].G
			dn[x].B += dn[x+1].B
		}
	}
	return pi
}


mean/mean.go
// Copyright 2013 Sonia Keys.
// Licensed under MIT license.  See "license" file in this source tree.

// Mean is a simple color quantizer.
package mean

import (
	"image"
	"image/color"
	"math"

	"github.com/soniakeys/quant"
)

// Quantizer implements quant.Quantizer with a simple mean-based algorithm.
type Quantizer struct{}

var _ quant.Quantizer = Quantizer{}

// Image performs color quantization and returns a paletted image.
//
// Argument n is the desired number of colors.  Returned is a paletted
// image with no more than n colors.
func (Quantizer) Image(img image.Image, n int) *image.Paletted {
	if n > 256 {
		n = 256
	}
	qz := newQuantizer(img, n)
	if n > 1 {
		qz.cluster() // cluster pixels by color
	}
	return qz.paletted() // generate paletted image from clusters
}

func (Quantizer) Palette(img image.Image, n int) quant.Palette {
	qz := newQuantizer(img, n)
	if n > 1 {
		qz.cluster() // cluster pixels by color
	}
	return qz.palette()
}

type quantizer struct {
	img image.Image // original image
	cs  []cluster   // len(cs) is the desired number of colors
}

type point struct{ x, y int32 }

type cluster struct {
	px       []point // list of points in the cluster
	widestCh int     // w const identifying channel with widest value range
	min, max uint32  // min, max color values of widest channel
	volume   uint64  // color volume
	priority int     // early: population, late: population*volume
}

const ( // w const
	wr = iota
	wg
	wb
)

func newQuantizer(img image.Image, n int) *quantizer {
	if n < 1 {
		return &quantizer{img, nil}
	}
	// Make list of all pixels in image.
	b := img.Bounds()
	px := make([]point, (b.Max.X-b.Min.X)*(b.Max.Y-b.Min.Y))
	i := 0
	for y := b.Min.Y; y < b.Max.Y; y++ {
		for x := b.Min.X; x < b.Max.X; x++ {
			px[i].x = int32(x)
			px[i].y = int32(y)
			i++
		}
	}
	// Make clusters, populate first cluster with complete pixel list.
	cs := make([]cluster, n)
	cs[0].px = px
	return &quantizer{img, cs}
}

// Cluster by repeatedly splitting clusters in two stages.  For the first
// stage, prioritize by population and split tails off distribution of
// color channel with widest range.  For the second stage, prioritize by
// the product of population and color volume, and split at the mean of
// the color channel with widest range.  Terminate when the desired number
// of clusters has been populated or when clusters cannot be further split.
func (qz *quantizer) cluster() {
	cs := qz.cs
	half := len(cs) / 2
	// cx is index of new cluster, populated at start of loop here, but
	// not yet analyzed.
	cx := 0
	c := &cs[cx]
	for {
		qz.setPriority(c, cx < half) // compute statistics for new cluster
		// determine cluster to split, sx
		sx := -1
		var maxP int
		for x := 0; x <= cx; x++ {
			// rule is to consider only clusters with non-zero color volume
			// and then split cluster with highest priority.
			if c := &cs[x]; c.max > c.min && c.priority > maxP {
				maxP = c.priority
				sx = x
			}
		}
		// If no clusters have any color variation, mark the end of the
		// cluster list and quit early.
		if sx < 0 {
			qz.cs = qz.cs[:cx+1]
			break
		}
		s := &cs[sx]
		m := qz.cutValue(s, cx < half) // get where to split cluster
		// point to next cluster to populate
		cx++
		c = &cs[cx]
		// populate c by splitting s into c and s at value m
		qz.split(s, c, m)
		// Normal exit is when all clusters are populated.
		if cx == len(cs)-1 {
			break
		}
		if cx == half {
			// change priorities on existing clusters
			for x := 0; x < cx; x++ {
				cs[x].priority =
					int(uint64(cs[x].priority) * (cs[x].volume >> 16) >> 29)
			}
		}
		qz.setPriority(s, cx < half) // set priority for newly split s
	}
}

func (q *quantizer) setPriority(c *cluster, early bool) {
	// Find extents of color values in each channel.
	var maxR, maxG, maxB uint32
	minR := uint32(math.MaxUint32)
	minG := uint32(math.MaxUint32)
	minB := uint32(math.MaxUint32)
	for _, p := range c.px {
		r, g, b, _ := q.img.At(int(p.x), int(p.y)).RGBA()
		if r < minR {
			minR = r
		}
		if r > maxR {
			maxR = r
		}
		if g < minG {
			minG = g
		}
		if g > maxG {
			maxG = g
		}
		if b < minB {
			minB = b
		}
		if b > maxB {
			maxB = b
		}
	}
	// See which channel had the widest range.
	w := wg
	min := minG
	max := maxG
	if maxR-minR > max-min {
		w = wr
		min = minR
		max = maxR
	}
	if maxB-minB > max-min {
		w = wb
		min = minB
		max = maxB
	}
	// store statistics
	c.widestCh = w
	c.min = min
	c.max = max
	c.volume = uint64(maxR-minR) * uint64(maxG-minG) * uint64(maxB-minB)
	c.priority = len(c.px)
	if !early {
		c.priority = int(uint64(c.priority) * (c.volume >> 16) >> 29)
	}
}

func (q *quantizer) cutValue(c *cluster, early bool) uint32 {
	var sum uint64
	switch c.widestCh {
	case wr:
		for _, p := range c.px {
			r, _, _, _ := q.img.At(int(p.x), int(p.y)).RGBA()
			sum += uint64(r)
		}
	case wg:
		for _, p := range c.px {
			_, g, _, _ := q.img.At(int(p.x), int(p.y)).RGBA()
			sum += uint64(g)
		}
	case wb:
		for _, p := range c.px {
			_, _, b, _ := q.img.At(int(p.x), int(p.y)).RGBA()
			sum += uint64(b)
		}
	}
	mean := uint32(sum / uint64(len(c.px)))
	if early {
		// split in middle of longer tail rather than at mean
		if c.max-mean > mean-c.min {
			mean = (mean + c.max) / 2
		} else {
			mean = (mean + c.min) / 2
		}
	}
	return mean
}

func (q *quantizer) split(s, c *cluster, m uint32) {
	px := s.px
	var v uint32
	i := 0
	last := len(px) - 1
	for i <= last {
		// Get pixel value of appropriate channel.
		r, g, b, _ := q.img.At(int(px[i].x), int(px[i].y)).RGBA()
		switch s.widestCh {
		case wr:
			v = r
		case wg:
			v = g
		case wb:
			v = b
		}
		// Split into two non-empty parts at m.
		if v < m || m == s.min && v == m {
			i++
		} else {
			px[last], px[i] = px[i], px[last]
			last--
		}
	}
	// Split the pixel list.
	s.px = px[:i]
	c.px = px[i:]
}

func (qz *quantizer) paletted() *image.Paletted {
	cp := make(color.Palette, len(qz.cs))
	pi := image.NewPaletted(qz.img.Bounds(), cp)
	for i := range qz.cs {
		px := qz.cs[i].px
		// Average values in cluster to get palette color.
		var rsum, gsum, bsum int64
		for _, p := range px {
			r, g, b, _ := qz.img.At(int(p.x), int(p.y)).RGBA()
			rsum += int64(r)
			gsum += int64(g)
			bsum += int64(b)
		}
		n64 := int64(len(px) << 8)
		cp[i] = color.RGBA{
			uint8(rsum / n64),
			uint8(gsum / n64),
			uint8(bsum / n64),
			0xff,
		}
		// set image pixels
		for _, p := range px {
			pi.SetColorIndex(int(p.x), int(p.y), uint8(i))
		}
	}
	return pi
}

func (qz *quantizer) palette() quant.Palette {
	cp := make(color.Palette, len(qz.cs))
	for i := range qz.cs {
		px := qz.cs[i].px
		// Average values in cluster to get palette color.
		var rsum, gsum, bsum int64
		for _, p := range px {
			r, g, b, _ := qz.img.At(int(p.x), int(p.y)).RGBA()
			rsum += int64(r)
			gsum += int64(g)
			bsum += int64(b)
		}
		n64 := int64(len(px) << 8)
		cp[i] = color.RGBA{
			uint8(rsum / n64),
			uint8(gsum / n64),
			uint8(bsum / n64),
			0xff,
		}
	}
	return quant.LinearPalette{cp}
}


median/median.go
// Copyright 2013 Sonia Keys.
// Licensed under MIT license.  See "license" file in this source tree.

// Median implements basic median cut color quantization.
package median

import (
	"container/heap"
	"image"
	"image/color"
	"math"
	"sort"

    "github.com/soniakeys/quant"
)

type Quantizer struct{}

var _ quant.Quantizer = Quantizer{}

// Quantize implements median cut color quantization.
//
// Argument n is the desired number of colors.  Returned is a paletted
// image with no more than n colors.
func (Quantizer) Image(img image.Image, n int) *image.Paletted {
	if n > 256 {
		n = 256
	}
	qz := newQuantizer(img, n)
	if n > 1 {
		qz.cluster() // cluster pixels by color
	}
	return qz.paletted() // generate paletted image from clusters
}

func (Quantizer) Palette(img image.Image, n int) quant.Palette {
    qz := newQuantizer(img, n)
    if n > 1 {
        qz.cluster() // cluster pixels by color
    }
    return qz.palette()
}

type quantizer struct {
	img image.Image // original image
	cs  []cluster   // len(cs) is the desired number of colors
	ch  chValues    // buffer for computing median
}

type point struct{ x, y int32 }
type chValues []uint16
type queue []*cluster

type cluster struct {
	px       []point // list of points in the cluster
	widestCh int     // w const identifying channel with widest value range
}

const ( // w const
	wr = iota
	wg
	wb
)

func newQuantizer(img image.Image, nq int) *quantizer {
	if nq < 1 {
		return &quantizer{img: img}
	}
	b := img.Bounds()
	npx := (b.Max.X - b.Min.X) * (b.Max.Y - b.Min.Y)
	qz := &quantizer{
		img: img,
		ch:  make(chValues, npx),
		cs:  make([]cluster, nq),
	}
	// Populate initial cluster with all pixels from image.
	c := &qz.cs[0]
	px := make([]point, npx)
	c.px = px
	i := 0
	for y := b.Min.Y; y < b.Max.Y; y++ {
		for x := b.Min.X; x < b.Max.X; x++ {
			px[i].x = int32(x)
			px[i].y = int32(y)
			i++
		}
	}
	return qz
}

// Cluster by repeatedly splitting clusters.
// Use a heap as priority queue for picking clusters to split.
// The rule is to spilt the cluster with the most pixels.
// Terminate when the desired number of clusters has been populated
// or when clusters cannot be further split.
func (qz *quantizer) cluster() {
	pq := new(queue)
	// Initial cluster.  populated at this point, but not analyzed.
	c := &qz.cs[0]
	var m uint32
	for i := 1; ; {
		// Only enqueue clusters that can be split.
		if qz.setWidestChannel(c) {
			heap.Push(pq, c)
		}
		// If no clusters have any color variation, mark the end of the
		// cluster list and quit early.
		if len(*pq) == 0 {
			qz.cs = qz.cs[:i]
			break
		}
		s := heap.Pop(pq).(*cluster) // get cluster to split
		m = qz.medianCut(s)
		c = &qz.cs[i] // set c to new cluster
		i++
		qz.split(s, c, m) // split s into c and s at value m
		// Normal exit is when all clusters are populated.
		if i == len(qz.cs) {
			break
		}
		if qz.setWidestChannel(s) {
			heap.Push(pq, s) // return s to queue
		}
	}
}

func (q *quantizer) setWidestChannel(c *cluster) bool {
	// Find extents of color values in each channel.
	var maxR, maxG, maxB uint32
	minR := uint32(math.MaxUint32)
	minG := uint32(math.MaxUint32)
	minB := uint32(math.MaxUint32)
	for _, p := range c.px {
		r, g, b, _ := q.img.At(int(p.x), int(p.y)).RGBA()
		if r < minR {
			minR = r
		}
		if r > maxR {
			maxR = r
		}
		if g < minG {
			minG = g
		}
		if g > maxG {
			maxG = g
		}
		if b < minB {
			minB = b
		}
		if b > maxB {
			maxB = b
		}
	}
	// See which channel had the widest range.
	c.widestCh = wg
	min := minG
	max := maxG
	if maxR-minR > max-min {
		c.widestCh = wr
		min = minR
		max = maxR
	}
	if maxB-minB > max-min {
		c.widestCh = wb
		min = minB
		max = maxB
	}
	return max > min
}

// Arg c must have value range > 0 in channel c.widestCh.
// return value m is guararanteed to split cluster into two non-empty clusters
// by v < m where v is pixel value of channel c.Widest.
func (q *quantizer) medianCut(c *cluster) uint32 {
	px := c.px
	ch := q.ch[:len(px)]
	// Copy values from appropriate channel to buffer for computing median.
	switch c.widestCh {
	case wr:
		for i, p := range c.px {
			r, _, _, _ := q.img.At(int(p.x), int(p.y)).RGBA()
			ch[i] = uint16(r)
		}
	case wg:
		for i, p := range c.px {
			_, g, _, _ := q.img.At(int(p.x), int(p.y)).RGBA()
			ch[i] = uint16(g)
		}
	case wb:
		for i, p := range c.px {
			_, _, b, _ := q.img.At(int(p.x), int(p.y)).RGBA()
			ch[i] = uint16(b)
		}
	}
	// Find cut.
	sort.Sort(ch)
	m1 := len(ch) / 2 // median
	if ch[m1] != ch[m1-1] {
		return uint32(ch[m1])
	}
	m2 := m1
	// Dec m1 until element to left is different.
	for m1--; m1 > 0 && ch[m1] == ch[m1-1]; m1-- {
	}
	// Inc m2 until element to left is different.
	for m2++; m2 < len(ch) && ch[m2] == ch[m2-1]; m2++ {
	}
	// Return value that makes more equitable cut.
	if m1 > len(ch)-m2 {
		return uint32(ch[m1])
	}
	return uint32(ch[m2])
}

func (q *quantizer) split(s, c *cluster, m uint32) {
	px := s.px
	var v uint32
	i := 0
	last := len(px) - 1
	for i <= last {
		// Get pixel value of appropriate channel.
		r, g, b, _ := q.img.At(int(px[i].x), int(px[i].y)).RGBA()
		switch s.widestCh {
		case wr:
			v = r
		case wg:
			v = g
		case wb:
			v = b
		}
		// Split at m.
		if v < m {
			i++
		} else {
			px[last], px[i] = px[i], px[last]
			last--
		}
	}
	// Split the pixel list.
	s.px = px[:i]
	c.px = px[i:]
}

func (qz *quantizer) paletted() *image.Paletted {
	cp := make(color.Palette, len(qz.cs))
	pi := image.NewPaletted(qz.img.Bounds(), cp)
	for i := range qz.cs {
		px := qz.cs[i].px
		// Average values in cluster to get palette color.
		var rsum, gsum, bsum int64
		for _, p := range px {
			r, g, b, _ := qz.img.At(int(p.x), int(p.y)).RGBA()
			rsum += int64(r)
			gsum += int64(g)
			bsum += int64(b)
		}
		n64 := int64(len(px) << 8)
		cp[i] = color.RGBA{
			uint8(rsum / n64),
			uint8(gsum / n64),
			uint8(bsum / n64),
			0xff,
		}
		// Set image pixels.
		for _, p := range px {
			pi.SetColorIndex(int(p.x), int(p.y), uint8(i))
		}
	}
	return pi
}

func (qz *quantizer) palette() quant.Palette {
    cp := make(color.Palette, len(qz.cs))
    for i := range qz.cs {
        px := qz.cs[i].px
        // Average values in cluster to get palette color.
        var rsum, gsum, bsum int64
        for _, p := range px {
            r, g, b, _ := qz.img.At(int(p.x), int(p.y)).RGBA()
            rsum += int64(r)
            gsum += int64(g)
            bsum += int64(b)
        }
        n64 := int64(len(px) << 8)
        cp[i] = color.RGBA{
            uint8(rsum / n64),
            uint8(gsum / n64),
            uint8(bsum / n64),
            0xff,
        }
    }
    return quant.LinearPalette{cp}
}

// Implement sort.Interface for sort in median algorithm.
func (c chValues) Len() int           { return len(c) }
func (c chValues) Less(i, j int) bool { return c[i] < c[j] }
func (c chValues) Swap(i, j int)      { c[i], c[j] = c[j], c[i] }

// Implement heap.Interface for priority queue of clusters.
func (q queue) Len() int { return len(q) }

// Priority is number of pixels in cluster.
func (q queue) Less(i, j int) bool { return len(q[i].px) > len(q[j].px) }
func (q queue) Swap(i, j int) {
	q[i], q[j] = q[j], q[i]
}
func (pq *queue) Push(x interface{}) {
	c := x.(*cluster)
	*pq = append(*pq, c)
}
func (pq *queue) Pop() interface{} {
	q := *pq
	n := len(q) - 1
	c := q[n]
	*pq = q[:n]
	return c
}
